driver-core/src/main/com/mongodb/ConnectionString.java

```diff
@@ -663,7 +663,14 @@ private MongoCredential createCredentials(final Map<String, List<String>> option
             }
 
             if (key.equals("authmechanism")) {
-                mechanism = AuthenticationMechanism.fromMechanismName(value);
+                if (value.equals("MONGODB-CR")) {
+                    if (userName == null) {
+                        throw new IllegalArgumentException("username can not be null");
```

Comment:
Seems unusual to throw here but warn that MONGODB-CR is deprecated.

Reply:
Yeah, this was tricky.  The warning is because we can't deprecate things in the connection string, so we don't want existing connection strings to cause runtime errors.  But if they don't supply a user name, it's still an error.  This is consistent with the current behavior, but due to the API changes I had to refactor where that check is done (it used to happen in createMongoDBCRCredential)

Reply:
SGTM